### DISCUSSION TITLE
vim.buf.lsp.rename() getting called twice

### DISCUSSION THREAD

**Discussion by debashisbiswas** [neovim/neovim#24121] [Q&A]
Hi all, I've recently been getting a weird issue in which `vim.buf.lsp.rename()` seems to get called twice in some projects. I suspect it might be because I'm using multiple language servers (Angular project has `angularls` and `tsserver`), as it doesn't happen in a python project with just `pyright`, but I'm not sure how to debug it or move towards a fix.

When I say it gets called twice, this is what I mean:
1. call `vim.buf.lsp.rename()` using keybind
2. `vim.ui.input()` prompt shows up, my cursor moves to the prompt, and I edit to a new name
3. symbol is renamed, and my cursor moves back to my buffer
4. `vim.ui.input()` prompt shows up again, and my cursor moves back to the prompt
5. I must leave the prompt manually or by typing something again

For reference, my configuration is here:
https://github.com/debashisbiswas/.dotfiles/tree/master/nvim/.config/nvim

**Comment by WillEhrendreich** [comment]
hey, @debashisbiswas, I think you're on to exactly why it's happening, as the request is going to all the attached servers, AFAIK. 

I'm no where near an expert in all of this, but it seems to me that you could pick one of them that you would consider "secondary" (my guess is angularls but legit I never touch javascript or web anything, so you're going to have to make that call), and then in your setup for it you could disable the capability. 

note this isn't tested, you may have to set it differently, somehow.. 

```
        mason_lspconfig.setup_handlers({
          function(server_name)
            if server_name == "angularls" then
              local renameCapabilityFalse = {
                renameProvider = False,
              }
              capabilities = vim.tbl_deep_extend("force", capabilities, renameCapabilityFalse)
            end
            require("lspconfig")[server_name].setup({
              capabilities = capabilities,
              on_attach = on_attach,
              settings = servers[server_name],
            })
          end,
        }),

```

Alternatively, and what I might actually do instead, you might be able to only invoke the request under the condition it isn't angularls asking, in the on_attach function where you set your keymaps..

```

---called on every server that attaches to a buffer, sets keymaps and other whatnot
---@param client lsp.Client
---@param bufnr integer
local on_attach = function(client, bufnr)
  local nmap = function(keys, func, desc)
    if desc then
      desc = 'LSP: ' .. desc
    end

    vim.keymap.set('n', keys, func, { buffer = bufnr, desc = desc })
  end

  nmap('<leader>rn',
      function()
        if not client.name == "angularls" then 
          vim.lsp.buf.rename()
        end
      end, '[R]e[n]ame')
    -- the rest of your on_attach here --
end 

```
notice that instead of discarding the client param, we specify it, and use it to check the name.

I hope this helps!

**Reply by fingust** [reply]
Your way of disabling the rename provider did not work for me, however putting this in the `on_attach` function did solve the problem:
```lua
if client.name == "angularls" then
    client.server_capabilities.renameProvider = false
end
```

**Reply by JulienLecoq** [reply]
It does not seems to be a good solution since you're completely removing angular renaming there. Hence, if you rename a variable from your .ts files, it will not be rename in your templates (.html files).

The best would be to start by calling the rename capability of angularls, if this one fails or is not available (could happen if you try to rename something that angularls does not support), then call the next rename capability of the next active lsp client and so on.

**Reply by fingust** [reply]
You're right! I was actually wondering why that wasn't working for me and I just dealt with it manually as I don't really need it that often, but did not even realize this was it.

At this point I'm not familiar enough with lua or lsp configuration to develop a solution you've described. But I tried to disable the renameProvider of the `tsserver` instead of `angularls`. Did a quick test and it seems to work. Although I don't know yet if this breaks something somewhere else, again.

**Reply by debashisbiswas** [reply]
I don't do enough Angular development anymore for this to be an issue for me, but I've looked in it some more. Research dump:

Disabling `angularls` renaming capability entirely isn't great, for the reasons mentioned above.

In looking for a better solution, I tried to figure out how VS Code handles this. Turns out, VS Code faces this issue as well: https://github.com/microsoft/vscode/issues/115354

One proposed solution was to have the Angular language service include a plugin for the TypeScript language server. This plugin would have TypeScript "yield" to Angular when needed. More info here: https://github.com/angular/angular/pull/44696, and here's the plugin's description (relevant to this discussion) -

> When the user wants to rename a symbol in the ts file VSCode
will ask the rename providers for the answer in turn. If the
first extension returns the result, the VSCode will break the
loop and apply the result. If the first extension cannot rename
the symbol, VSCode will ask the second extension in the
list (built-in TS/JS extension, Angular LS extension, etc.).
In other words, VSCode takes the result from only one rename provider
and the order depends on registration timing, scoring.

> Because the built-in ts extension and Angular extension have the
same high score, if the built-in ts extension is the
first(depends on the time the extension was registered), the result
will be provided by the built-in extension. We want Angular to
provide it, so this plugin will delegate rename requests and reject
the request for the built-in ts server.

> The Angular LS only provides the rename info when working within
an Angular project. If we cannot locate Angular sources in the
project files, the built-in extension should provide the rename info.

> This plugin will apply to the built-in TS/JS extension and delegate
rename requests to the Angular LS. It provides the rename info only
when it is an Angular project. Otherwise, it will return info by the
default built-in ts server rename provider.

VS Code included this in their Angular extension in this pull request: https://github.com/angular/vscode-ng-language-service/pull/1589, but it seems that change was reverted in https://github.com/angular/vscode-ng-language-service/pull/1687 - seems it did not work as expected.

Though, it looks like coc included it and never reverted it: https://github.com/iamcco/coc-angular/commit/9ae151183b7de85a00aca2c9b4874995eb8069ad

**Reply by guoliang** [reply]
Any update on this. I have the same issue, and I'm also using `angularls` and `tsserver`.

**Comment by fightingdreamer** [comment]
Use any available lsp:
- [lsp_buf_rename_use_any](https://github.com/fightingdreamer/dotfiles/blob/54bb8b90b1741f58e02e1911cb6de73d48160247/lua/nv/lua/core/opts_lsp.lua#L52)

Use lsp from priority list or use any available lsp:
- [lsp_buf_rename_use_priority_or_any](https://github.com/fightingdreamer/dotfiles/blob/54bb8b90b1741f58e02e1911cb6de73d48160247/lua/nv/lua/core/opts_lsp.lua#L93)

Use lsp from priority list or aks user to select lsp:
- [lsp_buf_rename_use_priority_or_select](https://github.com/fightingdreamer/dotfiles/blob/54bb8b90b1741f58e02e1911cb6de73d48160247/lua/nv/lua/core/opts_lsp.lua#L101)

**Comment by NicolaiSattler** [comment]
In my case it was caused by installing a double languageserver for a specific language.  😁

**Comment by Maryannah** [comment]
Late to the party, but if it can ever help someone in need, I ended up doing this (+ bonus for organizing imports command)

```lua
-- With lazy.nvim, on the "neovim/nvim-lspconfig" plugin
config = function(_, opts)
  vim.api.nvim_create_autocmd("LspAttach", {
    group = groupmaker("lsp-attach"),
    desc = "Attach capabilities when buffer opens",
    callback = function(event)
      local client = vim.lsp.get_client_by_id(event.data.client_id)
      local filetype = vim.bo[event.buf].filetype

      if client and client.name == "ts_ls" and (filetype == "typescript" or filetype == "javascript") then
        -- Disable typescript renaming if angular language service is present
        if vim.lsp.get_clients({ bufnr = event.buf, name = "angularls" }) then
          client.server_capabilities.renameProvider = false
        end

        vim.api.nvim_buf_create_user_command(event.buf, "OrganizeImports", function()
          local params = {
            command = "_typescript.organizeImports",
            arguments = { vim.api.nvim_buf_get_name(0) },
          }

          local result = vim.lsp.buf_request_sync(0, "workspace/executeCommand", params, 1000)
          if not result then
            vim.notify("Organizing imports failed")
          end
        end, {})

        vim.keymap.set("n", "<leader>o", "<cmd>OrganizeImports<CR>", { desc = "Organize imports" })
      end
      -- ... Other configurations of my own
```

This disables the typescript renaming capabilities in favor of the angular ones, since Angular is more powerful in its own context (can rename in HTML files too)

**Comment by ajesquirell** [comment]
Unfortunately this still has the issue of other LSP functions giving duplicate results, like lsp_references for a given variable. Seems like there should be a better way to handle this kind of thing.

**Reply by ajesquirell** [reply]
Here is a way I found to fix all LSP duplicated functionality. In my case I am using angularls and typescript-tools.

```lua
-- With lazy.nvim, on the "neovim/nvim-lspconfig" plugin
vim.api.nvim_create_autocmd('LspAttach', {
  group = vim.api.nvim_create_augroup('angular_lsp_exclusive', { clear = true }),
  callback = function(event)
    local client = vim.lsp.get_client_by_id(event.data.client_id)
    if not client then return end

    -- Fix angularls and typescript-tools conflicts (eg renaming, lsp_references happens twice)
    if client.name == 'angularls' or client.name == 'typescript-tools' then
      -- Let the current LspAttach finish before reconciling clients
      vim.schedule(function()
        if vim.api.nvim_buf_is_valid(event.buf) then
          local angular = vim.lsp.get_clients {
            bufnr = event.buf,
            name = 'angularls',
          }

          if #angular == 0 then return end

          for _, ts_client in
            ipairs(vim.lsp.get_clients {
              bufnr = event.buf,
              name = 'typescript-tools',
            })
          do
            vim.lsp.buf_detach_client(event.buf, ts_client.id)
          end
        end
      end)
    end
  end,
})

```

**Reply by ajesquirell** [reply]
After writing this, I found that angularls is not really a superset of the LSP functionality from other typescript LSPs. When using the above code (which completely detaches typescript-tools when angularls is active), things like vim.lsp.buf.definition() does not work nearly as well if at all.

I think this issue has to be handled on a case-by-case basis. In my case, I found that keeping them both active, but specifying which functionality I want to run only through angularls works well:

```lua
local angular_only_methods = {
  ['textDocument/references'] = true,
  ['textDocument/rename'] = true,
  ['textDocument/prepareRename'] = true,
}

-- Inside the 'LspAttach' callback:
if client.name == 'typescript-tools' and not client._angular_method_filter then
  local original_supports_method = client.supports_method
  client.supports_method = function(self, method, bufnr)
    if bufnr and angular_only_methods[method] then
      local angular_clients = vim.lsp.get_clients {
        bufnr = bufnr,
        name = 'angularls',
      }

      if #angular_clients > 0 then return false end
    end

    return original_supports_method(self, method, bufnr)
  end

  client._angular_method_filter = true
end
```
